// @scaleflex/ui/core/upload-input/upload-input.component.js
import _extends from "@babel/runtime/helpers/extends";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["error", "buttonLabel", "buttonColor", "placeholder", "onChange", "ref"];
import React, { useState } from 'react';
import Button from '../button';
import StyledInput from '../input/input.styles';
import Styled from './upload-input.styles';
import { ButtonColor } from '../../utils/types';
var UploadInput = function UploadInput(_ref) {
  var error = _ref.error,
    _ref$buttonLabel = _ref.buttonLabel,
    buttonLabel = _ref$buttonLabel === void 0 ? 'Choose file' : _ref$buttonLabel,
    _ref$buttonColor = _ref.buttonColor,
    buttonColor = _ref$buttonColor === void 0 ? ButtonColor.Primary : _ref$buttonColor,
    _ref$placeholder = _ref.placeholder,
    placeholder = _ref$placeholder === void 0 ? 'No file chosen' : _ref$placeholder,
    _onChange = _ref.onChange,
    ref = _ref.ref,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useState = useState(''),
    _useState2 = _slicedToArray(_useState, 2),
    valueState = _useState2[0],
    setValueState = _useState2[1];
  return /*#__PURE__*/React.createElement(Styled.UploadInput, {
    ref: ref
  }, /*#__PURE__*/React.createElement(StyledInput.Input, {
    error: error,
    onClick: function onClick(_ref2) {
      var currentTarget = _ref2.currentTarget;
      var fileInputElem = currentTarget.querySelector('input[type=file]');
      if (fileInputElem) {
        fileInputElem.click();
      }
    }
  }, /*#__PURE__*/React.createElement(StyledInput.Base, {
    placeholder: placeholder,
    readOnly: true,
    value: valueState
  }), /*#__PURE__*/React.createElement(Button, {
    color: buttonColor
  }, buttonLabel), /*#__PURE__*/React.createElement(Styled.FileInput, _extends({}, rest, {
    type: "file",
    onChange: function onChange(event) {
      var files = event.target.files;
      if (files) {
        if (files.length === 0) {
          setValueState('');
        } else if (files.length === 1) {
          setValueState(files[0].name);
        } else if (files.length > 1) {
          setValueState("".concat(files.length, " files"));
        } else {
          setValueState(event.target.value);
        }
      }
      if (typeof _onChange === 'function') {
        _onChange(event);
      }
    }
  }))));
};
export default UploadInput;
