// @scaleflex/ui/core/tab/tab.component.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["icon", "label", "badge", "active", "disabled", "notification", "className", "size", "ref"];
import React from 'react';
import Dot from '@scaleflex/icons/dot';
import { Size } from './types';
import Badge from '../badge';
import Styled from './tab.styles';
var Tab = function Tab(_ref) {
  var _ref$icon = _ref.icon,
    icon = _ref$icon === void 0 ? false : _ref$icon,
    label = _ref.label,
    badge = _ref.badge,
    active = _ref.active,
    disabled = _ref.disabled,
    notification = _ref.notification,
    className = _ref.className,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? Size.Md : _ref$size,
    ref = _ref.ref,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement(Styled.Tab, _extends({
    ref: ref,
    size: size,
    icon: icon,
    disabled: disabled
  }, rest, {
    className: "".concat(className).concat(active ? ' active' : '')
  }), /*#__PURE__*/React.createElement(Styled.Icons, null, icon && /*#__PURE__*/React.createElement(Styled.Icon, {
    disabled: disabled,
    size: size
  }, icon), notification && !disabled && /*#__PURE__*/React.createElement(Styled.Dot, null, /*#__PURE__*/React.createElement(Dot, {
    color: notification
  }))), label && /*#__PURE__*/React.createElement(Styled.Label, null, label), badge && !disabled && /*#__PURE__*/React.createElement(Styled.Badge, null, /*#__PURE__*/React.createElement(Badge, {
    inline: true,
    size: 14,
    fontSize: 10,
    padding: "0 1px",
    badgeContent: badge
  })));
};
Tab.displayName = 'Tab';
export default Tab;
