// @scaleflex/ui/core/switcher-group/switcher-group.component.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["size", "label", "switcherProps", "SwitcherProps", "checked", "onChange", "readOnly", "disabled", "ref"];
import React from 'react';
import Switcher from '../switcher';
import Styled from './switcher-group.styles';
import { SwitcherSize } from '../switcher/switcher-size';
var SwitcherGroup = function SwitcherGroup(_ref) {
  var _ref$size = _ref.size,
    size = _ref$size === void 0 ? SwitcherSize.Sm : _ref$size,
    label = _ref.label,
    switcherProps = _ref.switcherProps,
    SwitcherProps = _ref.SwitcherProps,
    _ref$checked = _ref.checked,
    checked = _ref$checked === void 0 ? false : _ref$checked,
    onChange = _ref.onChange,
    readOnly = _ref.readOnly,
    disabled = _ref.disabled,
    ref = _ref.ref,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement(Styled.SwitcherGroup, _extends({
    disabled: disabled,
    ref: ref
  }, rest), /*#__PURE__*/React.createElement(Switcher, {
    size: size,
    SwitcherPropsData: SwitcherProps,
    switcherProps: switcherProps,
    readOnly: readOnly,
    disabled: disabled,
    onChange: onChange,
    checked: checked
  }), /*#__PURE__*/React.createElement(Styled.Label, {
    disabled: disabled,
    size: size
  }, label));
};
export default SwitcherGroup;
