// @scaleflex/ui/core/radio/radio.component.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["size", "checked", "onChange", "radioProps", "readOnly", "disabled", "ref"];
import React from 'react';
import Styled from './radio.styles';
import { RadioSize } from './size';
var Radio = function Radio(_ref) {
  var _ref$size = _ref.size,
    size = _ref$size === void 0 ? RadioSize.Sm : _ref$size,
    _ref$checked = _ref.checked,
    checked = _ref$checked === void 0 ? false : _ref$checked,
    onChange = _ref.onChange,
    radioProps = _ref.radioProps,
    readOnly = _ref.readOnly,
    disabled = _ref.disabled,
    ref = _ref.ref,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement(Styled.Radio, _extends({
    disabled: disabled
  }, rest, {
    ref: ref
  }), /*#__PURE__*/React.createElement(Styled.Input, _extends({
    checked: checked,
    onChange: readOnly || disabled ? undefined : onChange,
    disabled: disabled,
    size: size
  }, radioProps)), /*#__PURE__*/React.createElement(Styled.RadioBox, {
    disabled: disabled,
    size: size
  }));
};
export default Radio;
