// @scaleflex/ui/core/autocomplete/autocomplete.styles.js
import styled, { css } from 'styled-components';
import CrossOutline from '@scaleflex/icons/cross-outline';
import StyledLabel from '../label/label.styles';
import StyledFormHint from '../form-hint/form-hint.styles';
import SfxMenu from '../menu';
import { generateClassNames, applyDisplayNames } from '../../utils/functions';
import { Color as PColor } from '../../utils/types/palette';
import StyledTag from '../tag/tag.styles';
import arrowTick from '../arrow-tick';
var baseClassName = 'Autocomplete';
var Autocomplete = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'root')
}).withConfig({
  componentId: "sc-uyc6rp-0"
})(["", "{margin-top:4px;}", "{margin-bottom:4px;}", "{max-width:100%;}", "{white-space:nowrap;overflow:hidden;text-overflow:ellipsis;display:block;}"], StyledFormHint.FormHint, StyledLabel.Label, StyledTag.Tag, StyledTag.Label);
var TickIcon = /*#__PURE__*/styled.span.attrs({
  className: generateClassNames(baseClassName, 'tickIcon')
}).withConfig({
  componentId: "sc-uyc6rp-1"
})(function (_ref) {
  var palette = _ref.theme.palette;
  return css(["color:", ";"], palette[PColor.AccentStateless]);
});
var CrossIcon = /*#__PURE__*/styled(CrossOutline).withConfig({
  componentId: "sc-uyc6rp-2"
})(["", ""], function (_ref2) {
  var palette = _ref2.theme.palette;
  return css(["color:", ";"], palette[PColor.IconsSecondary]);
});
var AutocompleteContainer = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'container')
}).withConfig({
  componentId: "sc-uyc6rp-3"
})([""]);
var InputIconEndContainer = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'inputIconEndContainer')
}).withConfig({
  componentId: "sc-uyc6rp-4"
})(["display:flex;align-items:center;gap:16px;"]);
var Menu = /*#__PURE__*/styled(SfxMenu).attrs({
  className: generateClassNames(baseClassName, 'menu')
}).withConfig({
  componentId: "sc-uyc6rp-5"
})(function (_ref3) {
  var maxMenuHeight = _ref3.maxMenuHeight,
    theme = _ref3.theme;
  return css(["max-height:", ";background:", ";"], +maxMenuHeight ? "".concat(+maxMenuHeight, "px") : maxMenuHeight, theme.palette[PColor.BackgroundStateless]);
});
var OptionGroup = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'optionGroup')
}).withConfig({
  componentId: "sc-uyc6rp-6"
})(function (_ref4) {
  var theme = _ref4.theme;
  return css(["&:not(:last-child){padding-bottom:8px;margin-bottom:8px;border-bottom:1px solid ", ";}"], theme.palette[PColor.BordersSecondary]);
});
var Arrow = /*#__PURE__*/styled(arrowTick).withConfig({
  componentId: "sc-uyc6rp-7"
})(["margin-left:auto;cursor:pointer;"]);
var Styled = applyDisplayNames({
  Autocomplete: Autocomplete,
  AutocompleteContainer: AutocompleteContainer,
  TickIcon: TickIcon,
  CrossIcon: CrossIcon,
  InputIconEndContainer: InputIconEndContainer,
  Menu: Menu,
  OptionGroup: OptionGroup,
  Arrow: Arrow
});
export default Styled;
