// @scaleflex/ui/core/switcher/switcher.component.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["size", "checked", "onChange", "SwitcherPropsData", "switcherProps", "readOnly", "disabled", "children", "ref"];
import React from 'react';
import { ignoreEvent } from '../../utils/functions';
import Styled from './switcher.styles';
import { SwitcherSize } from './switcher-size';
var Switcher = function Switcher(_ref) {
  var _ref$size = _ref.size,
    size = _ref$size === void 0 ? SwitcherSize.Sm : _ref$size,
    _ref$checked = _ref.checked,
    checked = _ref$checked === void 0 ? false : _ref$checked,
    onChange = _ref.onChange,
    SwitcherPropsData = _ref.SwitcherPropsData,
    switcherProps = _ref.switcherProps,
    readOnly = _ref.readOnly,
    disabled = _ref.disabled,
    children = _ref.children,
    ref = _ref.ref,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement(Styled.SwitcherWrapper, _extends({
    disabled: disabled,
    ref: ref
  }, SwitcherPropsData, rest), /*#__PURE__*/React.createElement(Styled.Input, _extends({
    checked: checked,
    onChange: readOnly || disabled ? ignoreEvent : onChange,
    readOnly: readOnly,
    $disabled: disabled,
    size: size
  }, switcherProps)), /*#__PURE__*/React.createElement(Styled.Switcher, {
    disabled: disabled,
    size: size
  }, children && /*#__PURE__*/React.createElement(Styled.SwitcherContentWrapper, {
    checked: checked,
    size: size
  }, children)));
};
export default Switcher;
