// @scaleflex/ui/core/ellipsed-text/ellipsed-text.styles.js
import styled, { css } from 'styled-components';
import { generateClassNames, applyDisplayNames } from '../../utils/functions';
var baseClassName = 'Ellipsed';
var EllipsedTextContainer = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'root')
}).withConfig({
  componentId: "sc-53xpj8-0"
})(["display:flex;"]);
var EllipsedTextWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'ellipsed-text-wrapper')
}).withConfig({
  componentId: "sc-53xpj8-1"
})(function (_ref) {
  var $maxLinesCount = _ref.$maxLinesCount;
  return css(["display:-webkit-box;-webkit-box-orient:vertical;-webkit-line-clamp:", ";overflow:hidden;width:100%;", ""], $maxLinesCount, $maxLinesCount === 1 && css(["word-break:break-all;"]));
});
var TooltipContent = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'tooltip-content')
}).withConfig({
  componentId: "sc-53xpj8-2"
})(function (_ref2) {
  var $customMaxHeight = _ref2.$customMaxHeight;
  return css(["max-height:", ";"], $customMaxHeight ? "".concat($customMaxHeight, "px") : '100%');
});
var SuffixTextWrapper = /*#__PURE__*/styled.div.attrs({
  className: generateClassNames(baseClassName, 'suffix-text-wrapper')
}).withConfig({
  componentId: "sc-53xpj8-3"
})(["word-break:normal;display:flex;justify-content:flex-end;flex-direction:column;"]);
var Styled = applyDisplayNames({
  EllipsedTextContainer: EllipsedTextContainer,
  EllipsedTextWrapper: EllipsedTextWrapper,
  TooltipContent: TooltipContent,
  SuffixTextWrapper: SuffixTextWrapper
});
export default Styled;
