// @scaleflex/ui/core/search/search.component.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["onChange", "value", "size", "ref"];
import React from 'react';
import SearchIcon from '@scaleflex/icons/search';
import { InputSize } from '../../utils/types';
import CrossButton from '../cross-button/cross-button.component';
import Input from '../input';
import Styled from './search.styles';
var Search = function Search(_ref) {
  var onChange = _ref.onChange,
    _ref$value = _ref.value,
    value = _ref$value === void 0 ? '' : _ref$value,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? InputSize.Md : _ref$size,
    ref = _ref.ref,
    rest = _objectWithoutProperties(_ref, _excluded);
  var handleClearIconClick = function handleClearIconClick(event) {
    if (onChange) {
      onChange(event);
    }
  };
  var getSearchIconSize = function getSearchIconSize(sizeName) {
    switch (sizeName) {
      case InputSize.Sm:
        return 16;
      case InputSize.Md:
      default:
        return 18;
    }
  };
  return /*#__PURE__*/React.createElement(Styled.Search, {
    size: size
  }, /*#__PURE__*/React.createElement(Input, _extends({
    ref: ref,
    value: value,
    size: size
  }, rest, {
    onChange: onChange,
    iconStart: function iconStart(iconProps) {
      return /*#__PURE__*/React.createElement(SearchIcon, _extends({}, iconProps, {
        size: getSearchIconSize(size)
      }));
    },
    clearIcon: (value === null || value === void 0 ? void 0 : value.length) !== 0 ? /*#__PURE__*/React.createElement(CrossButton, {
      size: size
    }) : undefined,
    clearIconClick: handleClearIconClick
  })));
};
export default Search;
