// @scaleflex/ui/core/notification/notification.component.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["status", "removeBackground", "hideIcon", "title", "message", "onClose", "ref"];
import React from 'react';
import PopupStatus from '../popup-status';
import CrossButton from '../cross-button';
import { NotificationStatus } from './types';
import Styled from './notification.styles';
var Notification = function Notification(_ref) {
  var _ref$status = _ref.status,
    status = _ref$status === void 0 ? NotificationStatus.Info : _ref$status,
    _ref$removeBackground = _ref.removeBackground,
    removeBackground = _ref$removeBackground === void 0 ? false : _ref$removeBackground,
    _ref$hideIcon = _ref.hideIcon,
    hideIcon = _ref$hideIcon === void 0 ? false : _ref$hideIcon,
    title = _ref.title,
    message = _ref.message,
    onClose = _ref.onClose,
    ref = _ref.ref,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement(Styled.NotificationWrapper, _extends({
    ref: ref,
    status: status,
    removeBackground: removeBackground,
    hideIcon: hideIcon,
    isTitle: Boolean(title)
  }, rest), /*#__PURE__*/React.createElement(Styled.Notification, null, !hideIcon && /*#__PURE__*/React.createElement(Styled.Icon, {
    removeBackground: removeBackground
  }, /*#__PURE__*/React.createElement(PopupStatus, {
    status: status,
    notificationBackground: !removeBackground
  })), /*#__PURE__*/React.createElement(Styled.MessageWrapper, null, title && /*#__PURE__*/React.createElement(Styled.Title, null, title), /*#__PURE__*/React.createElement(Styled.Message, null, message))), !removeBackground && /*#__PURE__*/React.createElement(Styled.Close, {
    onClick: onClose
  }, /*#__PURE__*/React.createElement(CrossButton, {
    size: "sm"
  })));
};
export default Notification;
